### ISSUE TITLE
Base rule extension: [no-constant-binary-condition] Remove the parts of the rule that TypeScript also reports on

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#9973]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+base+rule+extension%22) and found none that match my proposal.
- [X] I have searched the [current extension rule list](https://typescript-eslint.io/rules/?=extension#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the base rule

https://eslint.org/docs/latest/rules/no-constant-binary-expression

### Description

Coming over from https://devblogs.microsoft.com/typescript/announcing-typescript-5-6/#disallowed-nullish-and-truthy-checks1 -> #9653: TypeScript now includes a _"Disallowed Nullish and Truthy Checks"_ feature. The blog post says:

> Similar results can be achieved by enabling the ESLint `no-constant-binary-expression` rule, and you can [see some of the results they achieved in their blog post](https://eslint.org/blog/2022/07/interesting-bugs-caught-by-no-constant-binary-expression/); but the new checks TypeScript performs does not have perfect overlap with the ESLint rule, and we also believe there is a lot of value in having these checks built into TypeScript itself.

Shall we write an extension rule that only reports on what TypeScript doesn't?

### Fail

```TypeScript
if (true || inDebuggingOrDevelopmentEnvironment()) {
    // ...
}
```

### Pass

```TypeScript
if (true && inDebuggingOrDevelopmentEnvironment()) {
    // ...
}
```

### Additional Info

💖

**Comment by kirkwaiblinger** [comment]
My immediate question when I read those release notes, and my question for this issue, is - how does `no-unnecessary-condition` play into this? That currently flags both the pass and fail examples 🤔

**Comment by JoshuaKGoldberg** [comment]
That is a great question. _Are_ there any cases that `no-unnecessary-condition` won't catch but `no-constant-binary-condition` would?

**Comment by kirkwaiblinger** [comment]
Idk, that's my question 🤷‍♂️😂

**Comment by kirkwaiblinger** [comment]
Also, is the TS design pure syntactic analysis, or does it take into account types? 🤷‍♂️

**Comment by JoshuaKGoldberg** [comment]
I poked around a bit. It's >=99% covered by `no-unnecessary-condition`. Anything not currently covered by `no-unnecessary-condition`, we should consider a separate bug report or feature request for that rule.

For now, we can just turn off `no-constant-binary-condition` in our core ESLint configs (`eslint-recommended`).

**Comment by bradzacher** [comment]
I would tend to agree - I don't think there's any great rush. From experience the errors that these checks reveal are very much few and far between. They're real bugs but they don't occur often enough for us to rush out a disable to deduplicate the errors. And anyways - double reporting on these things ain't bad cos they are real, provable bugs in the code.

**Comment by JoshuaKGoldberg** [comment]
👍 closing for now. We can revisit in a year or two once TS<5.5 is much less common.
